File: wol.lua



wol_cmd	= "python ".. bot_path .."wol.py"


-- WOL 대상 MAC 주소, 각 시스템의 MAC Address 로 변경해서 사용한다.
wol_system	= 	{	
					[1] = '1A:2B:3C:4D:5E:6F', 	-- 데스크탑
					[2] = '00:AA:BB:CC:DD:EE'	-- 서버
				}

function WakeOnLan(user_id, system_idx)
	print('MAC : ' .. wol_system[system_idx] .. '시스템 WOL 요청')

	-- 192.168.0.255 는 사용하는 네트워크의 Broadcast 주소로 변경해서 사용한다.
	ExecuteCommand(wol_cmd .. ' -b 192.168.0.255 ' .. wol_system[system_idx])

	send_msg(user_id, 'MAC : ' .. wol_system[system_idx] .. '시스템 WOL 요청 완료', ok_cb, false)
end

File: wol.py
#!/usr/bin/env python
#
# A simple command line magic packet sender for WOL enabled devices
# Copyright (c) 2013, Petros Kyladitis <http://www.multipetros.gr/>
#
# This is free software distributed under the terms of FreeBSD license
      
      
        

import socket
import re
import argparse


def wol(macAddress, broadcastAddress="192.168.1.255", port=9):
    soc = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    soc.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)
    sendBytes = soc.sendto("\xff"*6 + GetMacHex(macAddress)*16, (broadcastAddress, port))
    return sendBytes


def GetMacHex(macAddress, separator=":"):
    if(macAddress.find(separator) == -1):
        separator = "-"
    macHex = macAddress.lower().replace(separator, "")
    if(re.match("[0-9a-f]{12}$", macHex)):
        return macHex.decode("hex")
    else:
        raise Exception("MAC address is not valid")


def main():
    parser = argparse.ArgumentParser(description="Send WOL magic packet to the selected MAC destination", epilog="(c) 2013, Petros Kyladitis. This is free software distributed under the terms of FreeBSD license. For updates and more info see at <http://www.multipetros.gr/>")
    parser.add_argument("mac", help="The MAC address of the machine you want to send the magic packet. Digits could be separated by \':\' or \'-\'", metavar="MAC_ADDRESS")
    parser.add_argument("-b", "--broadcast", default="192.168.1.255", dest='broadcast', help="The network broadcast address, in dotted format: www.xxx.yyy.zzz", metavar="BROADCAST_ADDRESS")
    parser.add_argument("-p", "--port", default="9", dest="port", help="The listening port usually 7 or 9. (By default, the program use port 9)", metavar="PORT_NUM")
    args = parser.parse_args()
    try:
        port = int(args.port)
        if((port < 1) or(port>65535)):
            raise Exception
    except Exception:
        print("Port must be a valid integer between 1-65535")
        return
    try:
        sendBytes = wol(args.mac, args.broadcast, port)
        print("Done! " + str(sendBytes) + " bytes send")
    except socket.gaierror:
        print(args.broadcast + " is not a valid broadcast address")
    except Exception as err:
        print(err)


if __name__ == "__main__":
    main()
